warehouse/oidc/models/activestate.py

```diff
@@ -0,0 +1,204 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import urllib
+
+from typing import Any
+
+from sqlalchemy import ForeignKey, String, UniqueConstraint
+from sqlalchemy.dialects.postgresql import UUID
+from sqlalchemy.orm import Query, mapped_column
+
+import warehouse.oidc.models._core as oidccore
+
+from warehouse.oidc.errors import InvalidPublisherError
+from warehouse.oidc.interfaces import SignedClaims
+from warehouse.oidc.models._core import (
+    CheckClaimCallable,
+    OIDCPublisher,
+    PendingOIDCPublisher,
+)
+
+ACTIVESTATE_URL = "https://platform.activestate.com"
+
+ACTIVESTATE_OIDC_API_PATH = "api/v1/oauth/oidc"
+
+
+def _check_sub(
+    ground_truth: str, signed_claim: str, _all_signed_claims: SignedClaims
+) -> bool:
+    # We expect a string formatted as follows:
+    #  org:<orgName>:project:<projectName>
+    # where :OPTIONAL-STUFF is a concatenation of other job context
+    # metadata. We currently lack the ground context to verify that
+    # additional metadata, so we limit our verification to just the ORG/REPO
+    # component.
+
+    # Defensive: ActiveState should never give us an empty subject.
+    if not signed_claim:
+        raise InvalidPublisherError("Missing 'subject' claim")
+
+    components = signed_claim.split(":")
+
+    if len(components) < 4:
+        raise InvalidPublisherError("Invalid 'subject' claim. Wrong format")
+
+    matches = (
+        f"{components[0]}:{components[1]}:{components[2]}:{components[3]}"
+        == ground_truth
+    )
+    if not matches:
+        raise InvalidPublisherError("Invalid 'subject' claim")
+
+    return True
+
+
+class ActiveStatePublisherMixin:
+    """
+    Common functionality for both pending and concrete ActiveState OIDC publishers.
+    """
+
+    organization = mapped_column(String, nullable=False)
+    activestate_project_name = mapped_column(String, nullable=False)
+    actor = mapped_column(String, nullable=False)
+    # 'actor' (The ActiveState platform username) is obstained from the user
+    # while configuring the publisher We'll make an api call to ActiveState to
+    # get the 'actor_id'
+    actor_id = mapped_column(String, nullable=False)
+    ingredient = mapped_column(String, nullable=True)
+
+    __required_verifiable_claims__: dict[str, CheckClaimCallable[Any]] = {
+        "sub": _check_sub,
+        "organization": oidccore.check_claim_binary(str.__eq__),
+        "project": oidccore.check_claim_binary(str.__eq__),
+        "actor_id": oidccore.check_claim_binary(str.__eq__),
+        # This is the name of the builder in the ActiveState Platform that
+        # publishes things to PyPI.
+        "builder": oidccore.check_claim_invariant("pypi-publisher"),
+    }
+
+    __optional_verifiable_claims__: dict[str, CheckClaimCallable[Any]] = {
+        # Should we check that the pypi projectname matches this ingredient
+        # name?
```

Comment:
What is an "ingredient" for an ActiveState project? Is it always the PyPI project name?

Reply:
The "ingredient" is the thing that is being built then published through our system.  It ultimately will be turned into the wheel that gets published.  I just checked with the team and it sounds like the ingredient name could potentially not match the PyPI project name but the wheel still be correct so I'm going to delete this comment.